File: odemapster-r2rml-rdb/src/main/java/es/upm/fi/dia/oeg/obdi/wrapper/r2rml/rdb/model/R2RMLSubjectMap.java
package es.upm.fi.dia.oeg.obdi.wrapper.r2rml.rdb.model;

import java.util.Collection;
import java.util.HashSet;

import org.apache.log4j.Logger;

import com.hp.hpl.jena.rdf.model.Resource;
import com.hp.hpl.jena.rdf.model.Statement;
import com.hp.hpl.jena.rdf.model.StmtIterator;

import es.upm.fi.dia.oeg.morph.base.Constants;
import es.upm.fi.dia.oeg.obdi.wrapper.r2rml.rdb.exception.R2RMLInvalidTermMapException;
import es.upm.fi.dia.oeg.obdi.wrapper.r2rml.rdb.model.R2RMLTermMap.TermMapType;

public class R2RMLSubjectMap extends R2RMLTermMap {
	
	private static Logger logger = Logger.getLogger(R2RMLSubjectMap.class);
	private Collection<String> classURIs;
	private R2RMLGraphMap graphMap;

//	public R2RMLSubjectMap(String constantValue) {
//		super(TermMapPosition.SUBJECT, constantValue);
//		super.setTermType(Constants.R2RML_LITERAL_URI());
//	}

	public static R2RMLSubjectMap create(String constantValue) {
		R2RMLSubjectMap sm = new R2RMLSubjectMap();
		sm.termMapType = TermMapType.CONSTANT;
		sm.constantValue = constantValue;
		sm.termType = Constants.R2RML_LITERAL_URI();
		return sm;
	}
	
	public static R2RMLSubjectMap create(Resource resource, R2RMLTriplesMap owner) 
			throws R2RMLInvalidTermMapException {
		R2RMLSubjectMap sm = new R2RMLSubjectMap();
		sm.parse(resource, owner);
		
		if(sm.getTermType() != null && sm.getTermType().equals(Constants.R2RML_LITERAL_URI())) {
			throw new R2RMLInvalidTermMapException("Literal is not permitted in the subject!");
		}
		
		StmtIterator classStatements = resource.listProperties(
				Constants.R2RML_CLASS_PROPERTY());
		if(classStatements != null) {
			sm.classURIs = new HashSet<String>();
			
			while(classStatements.hasNext()) {
				Statement classStatement = classStatements.nextStatement();
				sm.classURIs.add(classStatement.getObject().toString());
			}
		}
		
		Statement graphMapStatement = resource.getProperty(Constants.R2RML_GRAPHMAP_PROPERTY());
		if(graphMapStatement != null) {
			sm.graphMap = R2RMLGraphMap.create((Resource) graphMapStatement.getObject(), owner);
		}

		Statement graphStatement = resource.getProperty(Constants.R2RML_GRAPH_PROPERTY());
		if(graphStatement != null) {
			String graphStatementObjectValue = graphStatement.getObject().toString();
			if(!Constants.R2RML_DEFAULT_GRAPH_URI().equals(graphStatementObjectValue)) {
				sm.graphMap = R2RMLGraphMap.create(graphStatementObjectValue);
				logger.debug("this.graphMap = " + sm.graphMap);				
			}

		}
		
		return sm;

	}

	public Collection<String> getClassURIs() {
		return classURIs;
	}

	public R2RMLGraphMap getGraphMap() {
		return graphMap;
	}
	
}

File: morph-r2rml/src/main/scala/es/upm/dia/fi/oeg/morph/r2rml/model/R2RMLSubjectMap.scala
package es.upm.dia.fi.oeg.morph.r2rml.model

import scala.collection.JavaConversions._
import org.apache.log4j.Logger
import es.upm.fi.dia.oeg.morph.base.Constants
import com.hp.hpl.jena.rdf.model.Resource
import java.util.HashSet
import com.hp.hpl.jena.rdf.model.RDFNode

class R2RMLSubjectMap(termMapType:Constants.MorphTermMapType.Value
    , termType:Option[String], datatype:Option[String], languageTag:Option[String]
    , val classURIs:Set[String], val graphMaps:Set[R2RMLGraphMap] ) 
extends R2RMLTermMap(termMapType, termType, datatype, languageTag) {
	override val logger = Logger.getLogger(this.getClass().getName());
	
	val inferredTermType = this.inferTermType;
	if(inferredTermType != null && inferredTermType.equals(Constants.R2RML_LITERAL_URI)) {
		throw new Exception("Literal is not permitted in the subject map!");
	}
		
}

object R2RMLSubjectMap {
	val logger = Logger.getLogger(this.getClass().getName());
	
	def apply(rdfNode:RDFNode) : R2RMLSubjectMap = {
		val coreProperties = R2RMLTermMap.extractCoreProperties(rdfNode);
		//coreProperties = (termMapType, termType, datatype, languageTag)
		val termMapType = coreProperties._1;
		val termType = coreProperties._2;
		val datatype = coreProperties._3;
		val languageTag = coreProperties._4;
		
		val classURIs:Set[String] = rdfNode match {
		  case resourceNode:Resource => {
			  val classStatements = resourceNode.listProperties(Constants.R2RML_CLASS_PROPERTY);
			  val classURIsAux : Set[String]= if(classStatements != null) {
				  classStatements.map(classStatement => {
				    classStatement.getObject().toString();}).toSet;
			  } else {
				  Set.empty;
			  }
			  classURIsAux
		  }
		  case _ => { Set.empty }
		}
		
		val graphMaps:Set[R2RMLGraphMap] = rdfNode match {
		  case resourceNode:Resource => { R2RMLGraphMap.extractGraphMaps(resourceNode); }
		  case _ => {Set.empty}
		}

		val sm = new R2RMLSubjectMap(termMapType, termType, datatype
		    , languageTag, classURIs, graphMaps);
		sm.rdfNode = rdfNode;


		sm.parse(rdfNode)
		sm
	}
	

	def extractSubjectMaps(resource:Resource) : Set[R2RMLSubjectMap]= {
	  val tms = R2RMLTermMap.extractTermMaps(resource, Constants.MorphPOS.sub);
	  val result = tms.map(tm => tm.asInstanceOf[R2RMLSubjectMap]);
	  result;
	}

}